SRVKS-657: Add module for custom route labels and annotations (#34553)
openshift/openshift-docs: modules/serverless-customize-labels-annotations-routes.adoc

[id="serverless-customize-labels-annotations-routes_{context}"]
= Customizing labels and annotations for {product-title} routes

> {product-title} routes support the use of custom labels and annotations, which can be configured by modifying the `metadata` spec of a Knative service. Custom labels and annotations are propagated from the service to the Knative route, then to the Knative ingress, and finally to the {product-title} route.

Reviewer: ```suggestion
{product-title} routes support the use of custom labels and annotations, which you can configure by modifying the `metadata` spec of a Knative service. Custom labels and annotations are propagated from the service to the Knative route, then to the Knative ingress, and finally to the {product-title} route.
```

An attempt at making this more active.

Reviewer: Also, should "Knative ingress" be something like "Knative ingress gateway?"

Author: Waiting for a response from eng about the ingress thing

Reviewer 2: I think "Knative ingress" is better in this case.

In the sentence, the "Knative ingress" means the name of custom resource (`ingresses.networking.internal.knative.dev`) and we usually call it `knative ingress` or `kingres`, `king`.

.Prerequisites

* You must have the {ServerlessOperatorName} and Knative Serving installed on your {product-title} cluster.

.Procedure

. Create a Knative service that contains the label or annotation that you want to propagate to the {product-title} route:
+
.Example service YAML
[source,yaml]
----
apiVersion: serving.knative.dev/v1
kind: Service
metadata:
  name: <service_name>
  labels:
    <label_name>: <label_value>
  annotations:
    <annotation_name>: <annotation_value>
...
----
+

> .Example kn command

Reviewer: Would it make sense to break this up into two steps or move the service YAML into the prereqs? That way, this "create" step is only associated with the command.

Author: They are both ways to create though, that's kind of the point 🙂 you can either create a YAML object or use the kn command. Updating the .headings to make this clearer

Author: Eventually we'll have a better strategy around documenting CLI vs YAML vs ODC but I think that's out of scope for this PR at the minute, I'd prefer to just show both options here for now.

Reviewer: > They are both ways to create though, that's kind of the point 🙂 you can either create a YAML object or use the kn command. Updating the .headings to make this clearer

In that case, ISG asks for a different construction:

> **Handling nonsequential actions**
> When a procedure has mutually exclusive alternatives or any kind of branching condition, do not present the alternatives as sequential steps. Describe the branching condition in a single step, using a sublist or subprocedures to present the alternatives. Do not insert the word or between alternatives.

Author: Added these as a sublist instead (next commit), lmk if that's better 🙂 thanks

[source,terminal]
----
$ kn service create <service_name> \
  --image=<image> \
  --annotation <annotation_name>=<annotation_value> \
  --label <label_value>=<label_value>
----

> . Verify that the {product-title} route has been created with annotation or label that you added, by inspecting the output from the following command:

Reviewer: ```suggestion
. Verify that the {product-title} route has been created with the annotation or label that you added by inspecting the output from the following command:
```

Author: There's not really a good way to show it yet because we don't have a specific example of an annotation yet, the docs are generic on purpose. I'd say we could consider adding this if we go to flesh out the docs later and add more specific examples; I had tried to do this initially, but have been asked by QE not to because it was adding confusion for the moment.

[id="serverless-customize-labels-annotations-routes_{context}"]
= Customizing labels and annotations for {product-title} routes

{product-title} routes support the use of custom labels and annotations, which you can configure by modifying the `metadata` spec of a Knative service. Custom labels and annotations are propagated from the service to the Knative route, then to the Knative ingress, and finally to the {product-title} route.

.Prerequisites

* You must have the {ServerlessOperatorName} and Knative Serving installed on your {product-title} cluster.

.Procedure

> . Create a Knative service that contains the label or annotation that you want to propagate to the {product-title} route.

Reviewer: ISG asks for a comma here (see "Introducing the procedure"). 
```suggestion
. Create a Knative service that contains the label or annotation that you want to propagate to the {product-title} route:
```

> ** Create a service by using YAML:

Reviewer: Could be "To create a service by using YAML:"

Reviewer: Could it make sense to include a link to resource creation docs as an additional resource following this module?

Author: I think it's OK without this tbh, we have a ton of examples of creating resources in serverless docs and don't reference it every time, so I think if we were going to do this it would be better to make the decision to update it across the board. I will leave it off for now since it's a bit out of scope for this PR.

+
.Example service created by using YAML
[source,yaml]
----
apiVersion: serving.knative.dev/v1
kind: Service
metadata:
  name: <service_name>
  labels:
    <label_name>: <label_value>
  annotations:
    <annotation_name>: <annotation_value>
...
----

> ** Create a service by using the `kn` CLI:

Reviewer: Could be "To create a service by using the `kn` CLI:"

```suggestion
** To create a service by using the `kn` CLI, enter:
```

+
.Example service created by using a `kn` command
[source,terminal]
----
$ kn service create <service_name> \
  --image=<image> \
  --annotation <annotation_name>=<annotation_value> \
  --label <label_value>=<label_value>
----

> . Verify that the {product-title} route has been created with the annotation or label that you added by inspecting the output from the following command:

Reviewer: Is there anything in particular that I should look for in the output? I assume I'm looking for a route that has the annotation or label. Could note that right after the example command + callouts.

Author: Just that the label or annotation is there, I think this is OK as is since it'll depend what labels or annotations a user was adding.

[id="serverless-customize-labels-annotations-routes_{context}"]
= Customizing labels and annotations for {product-title} routes

{product-title} routes support the use of custom labels and annotations, which can be configured by modifying the `metadata` spec of a Knative service. Custom labels and annotations are propagated from the service to the Knative route, then to the Knative ingress, and finally to the {product-title} route.

The following annotations are supported by Knative Serving:

* Disabling routes (`DisableRouteAnnotation`), which is specified by the `serving.knative.openshift.io/disableRoute` annotation.

> * Enabling passthrough (`EnablePassthroughRouteAnnotation`), which is specified by the `serving.knative.openshift.io/enablePassthrough` annotation.

Reviewer 3: Not sure why these have their camelcased constant definitions here. Users would never see/know/use them.

Author: Literally just because I goofed wrt. the example and thought I should include them, I'd much rather take them out, thanks!

.Prerequisites

* You must have the {ServerlessOperatorName} and Knative Serving installed on your {product-title} cluster.

.Procedure

* Create a Knative service that contains the label or annotation that you want to propagate to the {product-title} route:
+
.Example service YAML
[source,yaml]
----
apiVersion: serving.knative.dev/v1
kind: Service
metadata:
  name: <service_name>
  labels:
    <label_name>: <label_value>
  annotations:
    <annotation_name>: <annotation_value>
...
----
+
.Example using DisableRouteAnnotation
[source,yaml]
----
apiVersion: serving.knative.dev/v1
kind: Service
metadata:

>   name: exxample-service

Reviewer 3: ```suggestion
  name: example-service
```

...
  annotations:

>     DisableRouteAnnotation: "serving.knative.openshift.io/disableRoute"

Reviewer 3: ```suggestion
    "serving.knative.openshift.io/disableRoute": "true"
```

Author: Oops, sorry, thanks

[id="serverless-customize-labels-annotations-routes_{context}"]
= Customizing labels and annotations for {product-title} routes

{product-title} routes support the use of custom labels and annotations, which can be configured by modifying the `metadata` spec of a Knative service. Custom labels and annotations are propagated from the service to the Knative route, then to the Knative ingress, and finally to the {product-title} route.

The following annotations are supported by Knative Serving:

* Disabling routes, which is specified by the `serving.knative.openshift.io/disableRoute` annotation.

> * Enabling pass-through, which is specified by the `serving.knative.openshift.io/enablePassthrough` annotation.

Reviewer 4: This general paragraph would fit better at the top of the page when we can enumerate all supported annotations like disableRoute, enablePassthrough or cluster local one.

Currently, this fragment is included as second one.

Author: I'll move it

> * Enabling pass-through routes, which is specified by the `serving.knative.openshift.io/enablePassthrough` annotation.

Reviewer 4: This section is misleading here. Previous paragraph says that any labels and annotations are passed. This paragraph list here some special annotations that have special effect, but that isn't full list (cluster local annotation is missing). Also, the description on `enablePassthrough` annotation is invalid - it should include information that this is about TLS termination passthough.

I think the best would be to remove this paragraph, or at least correct the description.

Author: OK. I was just adding these for the sake of completeness, when I asked Markus these were the Knative specific ones he mentioned, but I'll remove it and we can iterate on this another time if we need to.

.Prerequisites

* You must have the {ServerlessOperatorName} and Knative Serving installed on your {product-title} cluster.

.Procedure

. Create a Knative service that contains the label or annotation that you want to propagate to the {product-title} route:
+
.Example service YAML
[source,yaml]
----
apiVersion: serving.knative.dev/v1
kind: Service
metadata:
  name: <service_name>
  labels:
    <label_name>: <label_value>
  annotations:
    <annotation_name>: <annotation_value>
...
----
+

> .Example of disabling routes by using the annotation

Reviewer 4: The whole title for this section is: "Customizing labels and annotations for {product-title} routes"

Why we adding an example of "disabling routes by using the annotation"? Such example is already discussed in detail in include::modules/serverless-openshift-routes.adoc[leveloffset=+1].

I think what makes sense here is to showcase the passthrouth of those labels/annotations.

I've invoked the following command:

```bash
$ kn service create showcase \
  --image=quay.io/cardil/knative-serving-showcase \
  --annotation annot-a=b \
  --label label-c=d
```

And I see it did pass the label and annotation to a route:

```bash
$ kubectl get routes.route.openshift.io \
     -l serving.knative.openshift.io/ingressName=showcase \
     -l serving.knative.openshift.io/ingressNamespace=default \
     -n knative-serving-ingress -o yaml | egrep 'label-c|annot-a'
      annot-a: b
      label-c: d
            f:annot-a: {}
            f:label-c: {}
```

@nak3 @markusthoemmes Is that make sense to make this paragraph about passthrouth? If now, why the above section can't be used?

Author: I used this because it is a simple example - it doesn't matter if the example is shown elsewhere, or whether we show disable routes or passthrough, I was only including it to show how an annotation is added. I'll change it to passthrough.

> .Example of enabling pass-through routes by using the annotation
> [source,yaml]
> ----
> apiVersion: serving.knative.dev/v1
> kind: Service
> metadata:
>   name: example-service
> ...
>   annotations:
>     "serving.knative.openshift.io/enablePassthrough": "true"
> ...
> ----

Reviewer 4: Please use example of generic labels and annotations, like it was used in draft. Using this special annotation is misleading.

```suggestion
.Example of passing custom labels and annotations to route
[source,yaml]
----
apiVersion: serving.knative.dev/v1
kind: Service
metadata:
  name: example
  labels:
    staging: true
  annotations:
    my-annotation: some-value
...
----
```

Or just remove this example and stay with generic `<label_name>:<label_value>`, `<annotation_name>: <annotation_value>`. That's also okay.

+
.Example kn command
[source,terminal]
----
$ kn service create <service_name> \
  --image=<image> \
  --annotation <annotation_name>=<annotation_value> \
  --label <label_value>=<label_value>
----

. Verify that the {product-title} route has been created with annotation or label that you added, by inspecting the output from the following command:
+
.Example command for verification
[source,terminal]
----
$ oc get routes.route.openshift.io \
     -l serving.knative.openshift.io/ingressName=<service_name> \ <1>
     -l serving.knative.openshift.io/ingressNamespace=<service_namespace> \ <2>
     -n istio-system -o yaml \
         | grep -e "<label_name>: \"<label_value>\""  -e "<annotation_name>: <annotation_value>" <3>

> ----

Reviewer 4: We can't expect to have (grepping) routes if we are using `serving.knative.openshift.io/disableRoute` annotation.

Also, the `istio-system` here should be `knative-serving-ingress`.

Author: @nak3 or @markusthoemmes will need to respond to this one - if we update the previous example to passthrough, does that make this step OK?
